File: lib/sbi.c
#include <sbicall.h>

struct sbiret sbi_hart_start(size_t hartid, size_t start_addr, size_t opaque) {
  struct sbiret ret;
  asm volatile("mv a0, %1\n"
               "mv a1, %2\n"
               "mv a2, %3\n"
               "li a6, %4\n"
               "li a7, %5\n"
               "ecall\n"
               "mv %0, a0"
               : "=r"(ret.error)
               : "r"(hartid), "r"(start_addr), "r"(opaque),
                 "i"(SBI_EXT_HSM_HART_START), "i"(SBI_EXT_HSM)
               : "a0", "a1", "a2", "a7", "gp");
  return ret;
}

struct sbiret sbi_hart_stop(void) {
  struct sbiret ret;
  asm volatile("li a6, %1\n"
               "li a7, %2\n"
               "ecall\n"
               "mv %0, a0"
               : "=r"(ret.error)
               : "i"(SBI_EXT_HSM_HART_STOP), "i"(SBI_EXT_HSM)
               : "a0", "a7", "gp");
  return ret;
}

struct sbiret sbi_hart_suspend(size_t suspend_type, size_t resume_addr,
                               size_t opaque) {
  struct sbiret ret = {0, 0};
  return ret;
}

struct sbiret sbi_hart_wakeup(size_t hartid) {
  struct sbiret ret;
  asm volatile("mv a0, %1\n"
               "li a6, %2\n"
               "li a7, %3\n"
               "ecall\n"
               "mv %0, a0"
               : "=r"(ret.error)
               : "r"(hartid), "i"(SBI_EXT_HSM_HART_SUSPEND), "i"(SBI_EXT_HSM)
               : "a0", "a7", "gp");
  return ret;
}

struct sbiret inline sbi_set_timer(size_t stime_value) {
  struct sbiret ret = {0, 0};
  return ret;
}

struct sbiret sbi_hart_get_status(size_t hartid) {
  struct sbiret ret;
  asm volatile("mv a0, %1\n"
               "li a6, %2\n"
               "li a7, %3\n"
               "ecall\n"
               "mv %0, a0"
               : "=r"(ret.error)
               : "r"(hartid), "i"(SBI_EXT_HSM_HART_GET_STATUS), "i"(SBI_EXT_HSM)
               : "a0", "a7", "gp");
  return ret;
}

void sbi_shutdown(void) {
  asm volatile("li a7, %0\n"
               "ecall\n"
               :
               : "i"(SBI_SHUTDOWN));
}

File: kernel/arch/rv64/sbi.rs
use super::reg::sys::ecall;

pub struct SbiRet {
    error: i64,
}

// v0.1 legacy
pub const SBI_SET_TIMER: usize = 0;
pub const SBI_CONSOLE_PUTCHAR: usize = 1;
pub const SBI_CONSOLE_GETCHAR: usize = 2;
pub const SBI_CLEAR_IPI: usize = 3;
pub const SBI_SEND_IPI: usize = 4;
pub const SBI_REMOTE_FENCE_I: usize = 5;
pub const SBI_REMOTE_SFENCE_VMA: usize = 6;
pub const SBI_REMOTE_SFENCE_VMA_ASID: usize = 7;
pub const SBI_SHUTDOWN: usize = 8;

// v0.2 and higher
pub const SBI_EXT_BASE: usize = 0x10;
pub const SBI_EXT_TIME: usize = 0x54494D45;
pub const SBI_EXT_IPI: usize = 0x735049;
pub const SBI_EXT_RFENCE: usize = 0x52464E43;
pub const SBI_EXT_HSM: usize = 0x48534D;
pub const SBI_EXT_SRST: usize = 0x53525354;
pub const SBI_EXT_PMU: usize = 0x504D55;
pub const SBI_EXT_DBCN: usize = 0x4442434E;
pub const SBI_EXT_SUSP: usize = 0x53555350;
pub const SBI_EXT_CPPC: usize = 0x43505043;
pub const SBI_EXT_DBTR: usize = 0x44425452;
pub const SBI_EXT_SSE: usize = 0x535345;

pub const SBI_EXT_IPI_SEND_IPI: usize = 0x0;

/* SBI return error codes */
pub const SBI_SUCCESS: isize = 0;
pub const SBI_ERR_FAILED: isize = -1;
pub const SBI_ERR_NOT_SUPPORTED: isize = -2;
pub const SBI_ERR_INVALID_PARAM: isize = -3;
pub const SBI_ERR_DENIED: isize = -4;
pub const SBI_ERR_INVALID_ADDRESS: isize = -5;
pub const SBI_ERR_ALREADY_AVAILABLE: isize = -6;
pub const SBI_ERR_ALREADY_STARTED: isize = -7;
pub const SBI_ERR_ALREADY_STOPPED: isize = -8;
pub const SBI_ERR_NO_SHMEM: isize = -9;
pub const SBI_ERR_INVALID_STATE: isize = -10;
pub const SBI_ERR_BAD_RANGE: isize = -11;

// sbi functions for hsm extension
const SBI_EXT_HSM_HART_START: usize = 0x0;
const SBI_EXT_HSM_HART_STOP: usize = 0x1;
const SBI_EXT_HSM_HART_GET_STATUS: usize = 0x2;
const SBI_EXT_HSM_HART_SUSPEND: usize = 0x3;

pub fn sbi_hart_start(hartid: usize, start_addr: usize, opaque: usize) -> SbiRet {
    let error = ecall(hartid, start_addr, opaque, 0, 0, 0, SBI_EXT_HSM_HART_START) as i64;
    SbiRet { error }
}

pub fn sbi_hart_stop() -> SbiRet {
    let error = ecall(0, 0, 0, 0, 0, 0, SBI_EXT_HSM_HART_STOP) as i64;
    SbiRet { error }
}

pub fn sbi_hart_suspend(suspend_type: usize, resume_addr: usize, opaque: usize) -> SbiRet {
    // Assuming you will fill in the details later
    SbiRet { error: 0 }
}

pub fn sbi_hart_wakeup(hartid: usize) -> SbiRet {
    let error = ecall(hartid, 0, 0, 0, 0, 0, SBI_EXT_HSM_HART_SUSPEND) as i64;
    SbiRet { error }
}

pub fn sbi_send_ipi(hart_mask: usize, hart_mask_base: usize) -> SbiRet {
    let error = ecall(hart_mask, hart_mask_base, 0, 0, 0, 0, SBI_EXT_IPI_SEND_IPI) as i64;
    SbiRet { error }
}

pub fn sbi_set_timer(stime_value: usize) -> SbiRet {
    let error = ecall(stime_value, 0, 0, 0, 0, 0, SBI_SET_TIMER) as i64;
    SbiRet { error }
}

pub fn sbi_hart_get_status(hartid: usize) -> SbiRet {
    let error = ecall(hartid, 0, 0, 0, 0, 0, SBI_EXT_HSM_HART_GET_STATUS) as i64;
    SbiRet { error }
}

pub fn sbi_shutdown() -> !  {
    ecall(0, 0, 0, 0, 0, 0, SBI_SHUTDOWN) as i64;
    unreachable!("sbi_shutdown failed");
}

pub fn sbi_clear_ipi() {
    ecall(0, 0, 0, 0, 0, 0, SBI_CLEAR_IPI) as i64;
}
